### ISSUE TITLE
[JENKINS-8929] Deadlock when upstream and downstream jobs are blocked on each other

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-23#4327]
<p>Job A (deploy to functional test) is upstream of job B (run functional tests).</p>

<p>We need job A to block while its downstream job builds - otherwise a deployment while the functional tests are running will fail job B.</p>

<p>We need job B to block while its upstream job builds - otherwise trying to run the tests while the deployment is running will fail.</p>

<p>Unfortunately, if you configure A to block on downstream builds and B to block on upstream builds and A and B are scheduled at the same time they both block each other and are queued indefinitely, requiring manual intervention to clear the deadlock.</p>

<p>In this situation job A should build and B should remain blocked until A completes.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mahoney.png" title="mahoney's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mahoney">mahoney</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929" target="_blank">Deadlock when upstream and downstream jobs are blocked on each other</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mahoney.png" title="mahoney's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mahoney">mahoney</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2011-04-08T22:38:39+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 0
<li><b>imported</b>: 2025-11-23
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Job A (deploy to functional test) is upstream of job B (run functional tests).</p>

<p>We need job A to block while its downstream job builds - otherwise a deployment while the functional tests are running will fail job B.</p>

<p>We need job B to block while its upstream job builds - otherwise trying to run the tests while the deployment is running will fail.</p>

<p>Unfortunately, if you configure A to block on downstream builds and B to block on upstream builds and A and B are scheduled at the same time they both block each other and are queued indefinitely, requiring manual intervention to clear the deadlock.</p>

<p>In this situation job A should build and B should remain blocked until A completes.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mahoney.png" title="mahoney's avatar" /> mahoney:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=146401&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-146401">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>(In the example above A and B need to be separate jobs - if a change is made to a test only, only B should run, there is no need to redeploy the application.  However, if the application changes it needs to be redeployed and the tests need to be run after that deploy.)</p>
</pre>
</details>

</ul>
</details>

<p>(In the example above A and B need to be separate jobs - if a change is made to a test only, only B should run, there is no need to redeploy the application.  However, if the application changes it needs to be redeployed and the tests need to be run after that deploy.)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mahoney.png" title="mahoney's avatar" /> mahoney:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=146402&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-146402">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For reference, I found this thread describing the same problem with a possible solution:
<a href="http://jenkins.361315.n4.nabble.com/Upstream-Downstream-Deadlock-td3307350.html" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins.361315.n4.nabble.com/Upstream-Downstream-Deadlock-td3307350.html</a></p>
</pre>
</details>

</ul>
</details>

<p>For reference, I found this thread describing the same problem with a possible solution:<br/>
<a href="http://jenkins.361315.n4.nabble.com/Upstream-Downstream-Deadlock-td3307350.html" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins.361315.n4.nabble.com/Upstream-Downstream-Deadlock-td3307350.html</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mahoney.png" title="mahoney's avatar" /> mahoney:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=146530&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-146530">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Fix made on my clone of jenkins and pull request made here:
<a href="https://github.com/jenkinsci/jenkins/pull/69" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/69</a></p>
</pre>
</details>

</ul>
</details>

<p>Fix made on my clone of jenkins and pull request made here:<br/>
<a href="https://github.com/jenkinsci/jenkins/pull/69" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/69</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dogfood.png" title="dogfood's avatar" /> dogfood:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=146673&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-146673">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/565/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #565</a>
     <span class="error">&#91;FIXED JENKINS-8929&#93;</span> Changed the definition of a building downstream project to one that is Building, Waiting, Pending or Buildable
<a href="https://issues.jenkins.io/browse/JENKINS-8929" title="Deadlock when upstream and downstream jobs are blocked on each other" class="issue-link" data-issue-key="JENKINS-8929"><del>JENKINS-8929</del></a> follow up changes
recording <a href="https://issues.jenkins.io/browse/JENKINS-8929" title="Deadlock when upstream and downstream jobs are blocked on each other" class="issue-link" data-issue-key="JENKINS-8929"><del>JENKINS-8929</del></a> fix</p>

<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/jenkins/commit/b40e7b4228a6a5979fa484b2d2b15259bfab0ff5" class="external-link" target="_blank" rel="nofollow noopener">b40e7b4228a6a5979fa484b2d2b15259bfab0ff5</a>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/model/AbstractProject.java</li>
	<li>core/src/main/java/hudson/model/Queue.java</li>
</ul>


<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/jenkins/commit/86d88e5e7b341c884b156185051945678ce81ae8" class="external-link" target="_blank" rel="nofollow noopener">86d88e5e7b341c884b156185051945678ce81ae8</a>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/model/AbstractProject.java</li>
	<li>core/src/main/java/hudson/model/Queue.java</li>
</ul>


<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/jenkins/commit/05b9727edf1089b1c5e6e9461b4f62c71790065e" class="external-link" target="_blank" rel="nofollow noopener">05b9727edf1089b1c5e6e9461b4f62c71790065e</a>
Files : </p>
<ul>
	<li>changelog.html</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/565/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #565</a><br/>
     <span class="error">&#91;FIXED JENKINS-8929&#93;</span> Changed the definition of a building downstream project to one that is Building, Waiting, Pending or Buildable<br/>
<a href="https://issues.jenkins.io/browse/JENKINS-8929" title="Deadlock when upstream and downstream jobs are blocked on each other" class="issue-link" data-issue-key="JENKINS-8929"><del>JENKINS-8929</del></a> follow up changes<br/>
recording <a href="https://issues.jenkins.io/browse/JENKINS-8929" title="Deadlock when upstream and downstream jobs are blocked on each other" class="issue-link" data-issue-key="JENKINS-8929"><del>JENKINS-8929</del></a> fix</p>

<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/jenkins/commit/b40e7b4228a6a5979fa484b2d2b15259bfab0ff5" class="external-link" target="_blank" rel="nofollow noopener">b40e7b4228a6a5979fa484b2d2b15259bfab0ff5</a><br/>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/model/AbstractProject.java</li>
	<li>core/src/main/java/hudson/model/Queue.java</li>
</ul>


<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/jenkins/commit/86d88e5e7b341c884b156185051945678ce81ae8" class="external-link" target="_blank" rel="nofollow noopener">86d88e5e7b341c884b156185051945678ce81ae8</a><br/>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/model/AbstractProject.java</li>
	<li>core/src/main/java/hudson/model/Queue.java</li>
</ul>


<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/jenkins/commit/05b9727edf1089b1c5e6e9461b4f62c71790065e" class="external-link" target="_blank" rel="nofollow noopener">05b9727edf1089b1c5e6e9461b4f62c71790065e</a><br/>
Files : </p>
<ul>
	<li>changelog.html</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mahoney.png" title="mahoney's avatar" /> mahoney:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=146686&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-146686">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks - though I'm not totally convinced the change should be symmetric; I think there may be a case for blocking downstream builds when upstream builds are blocked, so that it is certain that if both are scheduled at the same time then it is definitely the upstream one that goes first.</p>
</pre>
</details>

</ul>
</details>

<p>Thanks - though I'm not totally convinced the change should be symmetric; I think there may be a case for blocking downstream builds when upstream builds are blocked, so that it is certain that if both are scheduled at the same time then it is definitely the upstream one that goes first.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cbos:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=147288&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147288">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We upgrade to version 1.403, since we also faced this issue.
But now the block does not work anymore in all situations.</p>

<p>We see situations were this fails. 
If both upstream and downstream job are queue and the downstream project finishes, both project get started.</p>

<p>Do I need to log a new issue for that?</p>

<p>As soon as a upstream project is running the downstream gets waiting in the queue.</p>
</pre>
</details>

</ul>
</details>

<p>We upgrade to version 1.403, since we also faced this issue.<br/>
But now the block does not work anymore in all situations.</p>

<p>We see situations were this fails. <br/>
If both upstream and downstream job are queue and the downstream project finishes, both project get started.</p>

<p>Do I need to log a new issue for that?</p>

<p>As soon as a upstream project is running the downstream gets waiting in the queue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mahoney.png" title="mahoney's avatar" /> mahoney:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=147385&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147385">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>My fault for not writing the test cases that would have shown Kohsuke why I did not make the change symmetrical.  I wanted to but I haven't managed to get my head round the testing strategy for Jenkins yet.</p>
</pre>
</details>

</ul>
</details>

<p>My fault for not writing the test cases that would have shown Kohsuke why I did not make the change symmetrical.  I wanted to but I haven't managed to get my head round the testing strategy for Jenkins yet.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=147654&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147654">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Robert Elliot
Path:
 core/src/main/java/hudson/model/AbstractProject.java
 core/src/main/java/hudson/model/Queue.java
<a href="http://jenkins-ci.org/commit/jenkins/bddc4f2c1ae569dbf7af4bc73f8dcf667bf407bb" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/bddc4f2c1ae569dbf7af4bc73f8dcf667bf407bb</a>
Log:
<span class="error">&#91;FIXED JENKINS-8929&#93;</span> Changed the definition of a building downstream project to one that is Building, Waiting, Pending or Buildable
but NOT blocked.  This prevents the deadlock condition where upstream project A is blocked on downstream project B, which is
blocked on upstream project A - now in this scenario B will remain blocked but A will build, eventually freeing up B.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Robert Elliot<br/>
Path:<br/>
 core/src/main/java/hudson/model/AbstractProject.java<br/>
 core/src/main/java/hudson/model/Queue.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/bddc4f2c1ae569dbf7af4bc73f8dcf667bf407bb" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/bddc4f2c1ae569dbf7af4bc73f8dcf667bf407bb</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-8929&#93;</span> Changed the definition of a building downstream project to one that is Building, Waiting, Pending or Buildable<br/>
but NOT blocked.  This prevents the deadlock condition where upstream project A is blocked on downstream project B, which is<br/>
blocked on upstream project A - now in this scenario B will remain blocked but A will build, eventually freeing up B.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=147655&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147655">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 core/src/main/java/hudson/model/AbstractProject.java
 core/src/main/java/hudson/model/Queue.java
<a href="http://jenkins-ci.org/commit/jenkins/1eff6776e81756b5f15c0b599210e433d366f11d" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/1eff6776e81756b5f15c0b599210e433d366f11d</a>
Log:
<a href="https://issues.jenkins.io/browse/JENKINS-8929" title="Deadlock when upstream and downstream jobs are blocked on each other" class="issue-link" data-issue-key="JENKINS-8929"><del>JENKINS-8929</del></a> follow up changes</p>

<ul class="alternate" type="square">
	<li>AbstractProject.getBuildingUpstream/getBuildingDownstream()
  should be symmetric.</li>
	<li>It's unlikely for us to add another category of items,
  so reverted getUnblockedItems() implementation to the older one
  that was more efficient.</li>
	<li>ItemList was meant to be internal, so keep it as such,
  at least for now.</li>
</ul>



</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 core/src/main/java/hudson/model/AbstractProject.java<br/>
 core/src/main/java/hudson/model/Queue.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/1eff6776e81756b5f15c0b599210e433d366f11d" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/1eff6776e81756b5f15c0b599210e433d366f11d</a><br/>
Log:<br/>
<a href="https://issues.jenkins.io/browse/JENKINS-8929" title="Deadlock when upstream and downstream jobs are blocked on each other" class="issue-link" data-issue-key="JENKINS-8929"><del>JENKINS-8929</del></a> follow up changes</p>

<ul class="alternate" type="square">
	<li>AbstractProject.getBuildingUpstream/getBuildingDownstream()<br/>
  should be symmetric.</li>
	<li>It's unlikely for us to add another category of items,<br/>
  so reverted getUnblockedItems() implementation to the older one<br/>
  that was more efficient.</li>
	<li>ItemList was meant to be internal, so keep it as such,<br/>
  at least for now.</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-8929?focusedId=147656&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147656">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 changelog.html
<a href="http://jenkins-ci.org/commit/jenkins/37f9f722fd15096066062931bfa264d8d8c31952" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/37f9f722fd15096066062931bfa264d8d8c31952</a>
Log:
recording <a href="https://issues.jenkins.io/browse/JENKINS-8929" title="Deadlock when upstream and downstream jobs are blocked on each other" class="issue-link" data-issue-key="JENKINS-8929"><del>JENKINS-8929</del></a> fix</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 changelog.html<br/>
<a href="http://jenkins-ci.org/commit/jenkins/37f9f722fd15096066062931bfa264d8d8c31952" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/37f9f722fd15096066062931bfa264d8d8c31952</a><br/>
Log:<br/>
recording <a href="https://issues.jenkins.io/browse/JENKINS-8929" title="Deadlock when upstream and downstream jobs are blocked on each other" class="issue-link" data-issue-key="JENKINS-8929"><del>JENKINS-8929</del></a> fix</p>
